PFW-1397 Add Toolchange counter

gudnimg opened the pull request:
This PR implements a new statistic for MMU2S users. Every tool change is now counted over the printer's lifetime. This statistic can be useful to determine when it is time to replace the "small" PTFE tube which is inside the extruder. 

The following will increment the counter:
1.  Toolchanges via T-code: `T0`, `T1`, `T2`, `T3`, `T4`, `T5`
2. Single material toolchange via T-code: `Tx` 
    - **NOTE:** `Tc` does not increment the counter. Because `Tx` and `Tc` are always used together, only `Tx` increments the counter to prevent double counts.
4. Any calls to `MMU2::load_filament_to_nozzle()`, this includes:
    - T-code `T?`
    - Loading to the nozzle via LCD menu
    - G-codes: `M600` and `M701` 

⚠️ **NOTE:**
> Currently it has not been decided how to show this statistic to users, so for the time being I've placed it inside the **Fail stats MMU** menu for convenience only. It had one unused menu item row.

----

Change in memory:
Flash: +262 bytes
SRAM: +1 bytes

NOTE: The EEPROM code is eating up a lot of this.

gudnimg committed:
PFW-1397 Allocate EEPROM memory for toolchange counter

gudnimg committed:
PFW-1397 Increment toolchange statistic

gudnimg committed:
PFW-1397 Add a temporary menu item to show the statistic

The location of the menu has not been decided. Currently I placed it where convenient.

gudnimg committed:
PFW-1397 Implement new function increment_tool_change_counter()

Reduces flash consumption by 44 bytes

leptun commented:
I'm not sure it's a good idea to update the eeprom for every toolchange. This would wear out the eeprom of the atmega2560 after 100000 toolchanges. This exact thing happened in the MMU firmware where there is a load balancing algorithm for the eeprom cells. It might be better if this statistic is only saved to eeprom at the end of the print. What do you think about this proposal?

gudnimg commented:
@leptun I like that idea, but that means we need to sacrifice 2 bytes of SRAM which I am fine with. 2 bytes because the print can easily go over 255 toolchanges, but I doubt any print goes over 65535 toolchanges. 

I also want to limit the EEPROM read within the menu. With the way I have it now, we are reading from the EEPROM address in a loop(). Might be better to either add a timer or find a way to have the value constant when the menu is opened.

leptun commented:
> find a way to have the value constant when the menu is opened.

You may be able to use the menu_data scratch buffer for this. Look at the SD menu and how the EEPROM_SD_SORT is cached

gudnimg committed:
PFW-1397 To reserve EEPROM write cycles, only update toolchange counter when a print finishes or is stopped

gudnimg committed:
PFW-1397 Implement ON_MENU_ENTER to simplify code

ON_MENU_ENTER runs code within the brackets only once
after a new menu is entered.

For the tool change menu, this allows us to display the data and
read from EEPROM only once.

gudnimg commented:
Note: Issue https://github.com/prusa3d/Prusa-Firmware/issues/3270 can now be fixed easily if we enclose all relevant `lcd_timeoutToStatus.stop()` calls with `ON_MENU_ENTER()`.

gudnimg committed:
PFW-1397 cleanup statistics menu

Change in memory:
Flash: -18 bytes
SRAM: -1 byte

gudnimg committed:
PFW-1397 minor comment tweaks

DRracer commented:
@gudnimg we can save the toolchanges per print after the print finishes just like other stats. We'd lose the stats only if the user resets the printer mid print (i.e. we can save the value after print finished or stopped).

Those guaranteed 100K EEPROM writes can be achieved pretty fast assuming one toolchange per minute:
100K/60/24 = ~69 days, which is something like 3 bigger prints :wink: 
On the MMU side the same amount of toolchanges is split across a wider EEPROM memory range.

gudnimg commented:
For what its worth I did print two prints this weekend to test the counter. And seems to work great. Counter is updated when the print finishes. 

![58882960-FB52-4B57-B0CB-58C7F0A83755](https://user-images.githubusercontent.com/8218499/197469494-5ca56607-70de-4907-964d-15ff312d218a.jpeg)

gudnimg commented on Firmware/menu.cpp:
```diff
@@ -347,6 +349,11 @@ bool __attribute__((noinline)) menu_item_leave(){
 
+bool menu_item_enter() {
+    menu_changed = false;
+    return !menu_changed;
```
I suspect this is not correct. This always returns true. Will look at this when I have time.

gudnimg committed:
PFW-1397 Fix: ON_MENU_ENTER was always run

gudnimg committed:
PFW-1397 Drop ON_MENU_ENTER

Using _menu_data_t turns out to save *just* a little bit more flash. So for now just use it.

gudnimg committed:
fixup

DRracer commented on Firmware/mmu2.h:
```diff
@@ -286,6 +300,7 @@ class MMU2 {
     
     bool inAutoRetry;
     uint8_t retryAttempts;
+    uint16_t toolchange_counter;
```
Total toolchange count is 32bit in eeprom while this variable is only 16bit. That assumes there will be no more than 64K toolchanges in a single print ... on MK3 this is probably a realistic assumption...
